saleor/discount/utils/voucher.py

```diff
@@ -525,12 +604,61 @@ def calculate_line_discount_amount_from_voucher(
     return discount_amount
 
 
+def calculate_order_line_discount_amount_from_denormalized_voucher(
+    line_info: "EditableOrderLineInfo", total_price: Money
+) -> Money:
+    """Calculate discount amount for line-level vouchers with denormalized values."""
+    voucher_info = line_info.voucher_denormalized_info
+    if not voucher_info:
+        return zero_money(total_price.currency)
+
+    quantity = line_info.line.quantity
+    currency = total_price.currency
+    if not voucher_info.apply_once_per_order:
```

Comment:
This should be splited into separate function.
One for handling voucher with `apply_once_per_order`, and one without.